design_notes/v2.md

```diff
@@ -0,0 +1,98 @@
+- [Use obstore / obspec for globbing in `open_virtual_mfdataset`](https://github.com/zarr-developers/VirtualiZarr/issues/569)
+- [Function definition for open_virtual_dataset in next major release](https://github.com/zarr-developers/VirtualiZarr/issues/553)
+- [Support passing configuration options to default_object_store](https://github.com/zarr-developers/VirtualiZarr/issues/559)
+- [Redefine Virtual Readers as func(filepath) -> ManifestStore](https://github.com/zarr-developers/VirtualiZarr/issues/498)
+- [Make xarray an optional dependency?](https://github.com/zarr-developers/VirtualiZarr/issues/521)
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: An object defining the paths, chunk byte offsets, lengths of one virtual array.
+- `ManifestArray`: A chunk manifest and its associated Zarr V3 ArrayMetadata
+- `ManifestGroup`: ManifestGroups are a way to organizing child groups and arrays into a data hierarchy. Contains Zarr `GroupMetadata` and one or more ManifestGroups and/or ManifestArrays.
+- `ManifestStore`: A ManifestGroup and its associated ObjectReader (see protocol below).
+- `xarray.Dataset`: Contains ManifestArray-backed variables and (optionally) in-memory NumPy-backed variables.
+
+## Definitions
+
+- `ObjectReader` - An implementation of the `GetAsync` and `GetRangeAsync` protocols defined in obspec. `ObjectReader`s are used to load arrays using `ManifestStore`.
+- `ReadableFile` - A reader that implements the `Read`, `Seek`, and `Tell` methods using a similar interface as a [Python BufferedReader](https://docs.python.org/3/library/io.html#io.BufferedReader). `ReadableFile`s are used by a `Backend` to construct a `ManifestStore`.
+- `Backend` - A backend is a function that constructs a `ManifestStore` from a `ReadableFile`, `ObjectReader`, and optionally `kwargs` specified using `virtual_backend_kwargs`.
+
+
+## Protocols
+
+```python
+from obspec import GetAsync, GetRangeAsync, Read, Seek, Tell
+
+class ObjectReader(GetAsync, GetRangeAsync):
+    ...
+
+class ReadableFile(Read, Seek, Tell):
+    ...
+```
+
+```python
+class Backend(protocol):
+    """
+    Callback protocol for creating a virtual zarr store from a file
+    """
+    def __call__(
+        url_string: str,
+        object_reader: ObjectReader,
```

Comment:
We could make object_reader optional if there are no loadable variables.

Reply:
> We could make object_reader optional if there are no loadable variables.

Well then it would be optional in general because we don't know if there are any `loadable_variables` at `ManifestStore` construction time.

Reply:
In the latest version, the `object_reader` also used by parsers to get the byte ranges for chunks rather than only to load data so I think it's definitely required in all cases.